### PR TITLE
fix(evolve): four correctness fixes in the evolve loop

### PR DESCRIPTION
**Pull request by amacbride**
## Summary

Four independent correctness fixes in the `evolve` agentic loop, one per commit, each with its own test. Stacked on current `develop`.

### 1. reset `build_verified` on edit and on failed build_check (`mod.rs`)
`build_verified` latched `true` on the first passing `build_check` and was never cleared. Two ways this let `done` complete an **unbuildable** config:
- build passes → agent edits again → `done` accepted, though the new edit was never built.
- build passes → a later build fails → `done` accepted as "complete" despite the failing build.

Now cleared whenever an edit is applied (`Edit` / `EditSemantic` / `EnsureSecret` arms) and whenever a `build_check` fails, so the done-gate requires a fresh, passing build.

### 2. handle `${ }` antiquotation in `validate_nix_syntax` (`file_ops.rs`)
The structural validator scanned strings with a flat `in_string` flag and never recognized `${ … }` antiquotation. A nested string inside an interpolation — e.g. `"${if a then "}" else "y"}"` — toggled the flag and mis-counted the braces, so the validator **rejected valid Nix and blocked legitimate agent edits**. Rewritten around a single context stack: `${` inside a string pushes a code frame, its matching `}` pops back. All prior behavior and error messages preserved.

### 3. don't duplicate a non-list attr in `edit_nix_file` `add()` (`edit_nix_file.rs`)
When `add()` targeted an attrpath that already existed as a **non-list** assignment (e.g. `programs.git = { … }`), it fell through and inserted a second `attrpath = [ … ];`, producing two assignments to the same attribute — **invalid Nix**. Now returns a clear error instead.

### 4. make `truncate_for_log` UTF-8 safe (`tools.rs`)
`truncate_for_log` sliced with `&s[..max_len]` (a byte index). A multi-byte char (accented letters, emoji, arrows — common in agent thoughts, truncated at 200 for logging) straddling that index **panicked** with "byte index N is not a char boundary." Now truncates on a char boundary, treating `max_len` as a character budget.

## Test Plan

- [ ] No test plan needed

Each commit adds its own test. Verified **red-green**: reverting each fix to develop's code (keeping the test) makes exactly that test fail — and only those tests, no collateral.

```
cargo test evolve::      # 166 passed / 0 failed
cargo clippy --tests     # clean for all four files
```

Tests added:
- Bug 1: `failing_build_check_clears_prior_verification`, `done_is_rejected_after_a_failing_build`, `edit_after_passing_build_clears_verification` (drive `process_tool_result` directly)
- Bug 2: accept/reject tests for nested-string and indented-string interpolation
- Bug 3: `add_refuses_to_duplicate_non_list_attr`
- Bug 4: short-string, newline-collapse, ASCII, and multi-byte-boundary cases

## Docs

- [ ] Docs updated (companion PR in darkmatter/nixmac-web: #\_\_\_)
- [x] No docs update needed

These are internal correctness fixes with no new user-facing behavior — they make already-valid configs work (2, 3) and stop a logging panic (4), and tighten an internal completion gate (1).

---
🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: apps/native/src-tauri/src/evolve/file_ops.rs [hunk 1]
```diff
@@ -174,21 +174,30 @@ fn path_scope_error(input: &str, base: &Path, resolved: &Path, operation: &str)
...
-    let mut in_block_comment = false;
-    let mut escape_next = false;
+    // Track structural balance (braces/brackets/parens) and string contexts.
+    // `${ ... }` antiquotation re-enters code context *inside* a string, so a
+    // single shared stack is used: an interpolation pushes a code frame and its
+    // matching `}` pops back to the string. This lets interpolations contain
+    // nested strings, braces, and further interpolations without desyncing the
+    // string state (which the old flag-based scanner got wrong, falsely
+    // rejecting valid Nix like `"${if a then "}" else "y"}"`).
+    #[derive(PartialEq)]
```
[the remark this answers is not in the document]
Comment (amacbride) [reply apps/native/src-tauri/src/evolve/file_ops.rs:184]: Declining: this compiles as-is (CI's Rust Check, Analyze (rust), and build are all green). `==` borrows its operands — `*f == Frame::Str` desugars to `eq(&*f, &Frame::Str)`, so there's no move out of the shared reference and `Copy` isn't required.

**Review by czxtm — APPROVED** [review]
